cortex-ar/src/mmu.rs

```diff
@@ -198,24 +267,51 @@ impl L1Section {
     /// # Panics
     ///
     /// Physcal address not aligned to 1 MB.
-    pub const fn new(phys_addr: u32, section_attrs: SectionAttributes) -> Self {
+    pub const fn new_with_addr_and_attrs(phys_addr: u32, section_attrs: SectionAttributes) -> Self {
         // Must be aligned to 1 MB
         if phys_addr & 0x000F_FFFF != 0 {
             panic!("physical base address for L1 section must be aligned to 1 MB");
         }
-        let higher_bits = phys_addr >> 20;
-        let raw = (higher_bits << 20)
-            | ((section_attrs.non_global as u32) << 17)
-            | ((section_attrs.shareable as u32) << 16)
-            | ((section_attrs.access.apx() as u32) << 15)
-            | ((section_attrs.memory_attrs.type_extensions as u32) << 12)
-            | ((section_attrs.access.ap() as u32) << 10)
-            | ((section_attrs.p_bit as u32) << 9)
-            | ((section_attrs.domain as u32) << 5)
-            | ((section_attrs.execute_never as u32) << 4)
-            | ((section_attrs.memory_attrs.c as u32) << 3)
-            | ((section_attrs.memory_attrs.b as u32) << 2)
-            | L1EntryType::Section as u32;
-        Self::new_with_raw_value(raw)
+        let attrs = section_attrs.l1_section_part();
+        L1Section::builder()
+            .with_base_addr_upper_bits(u12::new((phys_addr >> 20) as u16))
+            .with_ng(attrs.ng())
+            .with_s(attrs.s())
+            .with_apx(attrs.apx())
+            .with_tex(attrs.tex())
+            .with_ap(attrs.ap())
+            .with_p_bit(attrs.p_bit())
+            .with_domain(attrs.domain())
+            .with_xn(attrs.xn())
+            .with_c(attrs.c())
+            .with_b(attrs.b())
+            .with_entry_type(attrs.entry_type())
+            .build()
+    }
+
+    /// Retrieve the section attributes.
+    #[inline]
+    pub fn section_attrs(&self) -> Result<SectionAttributes, InvalidL1EntryType> {
+        SectionAttributes::from_raw(self.raw_value())
+    }
+
+    /// Set the section attributes without changing the address.
+    #[inline]
+    pub fn set_section_attrs(&mut self, section_attrs: SectionAttributes) {
+        let attrs = section_attrs.l1_section_part();
```

Comment:
Can you call the new_with_addrs_and_attrs function here, passing the existing address?

Reply:
That would perform checks which are not necessary anymore.

Reply:
Hmm, I could just add a `new_with_addrs_and_attrs_unchecked` method (a bit verbose, but I guess that's fine) and call that one.

Reply:
You could have `new_with_addr_and_attrs` call `set_section_attrs`